[REEF-1812] Avro bridge C# code generation

dougmsft opened the pull request:
- Adds a new project Org.Apache.REEF.Bridge.CLR that will be
    become Org.Apache.REEF.Bridge when the Java/C# interop bridge is removed.
  - Adds Avro avsc to C# code generation task to the msbuild project file.
  - Adds a number of Avro messages for Java/C# Avro bridge.

JIRA:
  [REEF-1812](https://issues.apache.org/jira/browse/REEF-1812)

Pull request:
  This closes #

markusweimer commented on ':
```diff
@@ -0,0 +1,178 @@
+  <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" />
+  <Import Project="$(SolutionDir)\.nuget\NuGet.targets" Condition="Exists('$(SolutionDir)\.nuget\NuGet.targets')" />
+  <Import Project="$(PackagesDir)\StyleCop.MSBuild.$(StyleCopVersion)\build\StyleCop.MSBuild.Targets" Condition="Exists('$(PackagesDir)\StyleCop.MSBuild.$(StyleCopVersion)\build\StyleCop.MSBuild.Targets')" />
+  <UsingTask TaskName="AvscToCSharp" TaskFactory="CodeTaskFactory" AssemblyFile="$(MSBuildToolsPath)\Microsoft.Build.Tasks.v4.0.dll">
```
Maybe move this task into its own `.tasks` file? We might want to use this code elsewhere in the future.

dougmsft replied:
Okay.

markusweimer commented on lang/common/bridge/avro/SystemOnStart.avsc:
```diff
@@ -0,0 +1,36 @@
+      {
+        "name":"dateTime",
+        "type":"long",
+        "doc":"Date time as a long since January 1, 1970"
```
What's the time unit? Seconds?

dougmsft replied:
Yes, it is seconds. I will update the comments.

markusweimer commented on lang/cs/Org.Apache.REEF.Bridge.CLR/Message/Acknowledgement.cs:
```diff
@@ -0,0 +1,66 @@
+// KIND, either express or implied.  See the License for the
+// specific language governing permissions and limitations
+// under the License.
+//<auto-generated />
```
Do we have to add the generated files for now?

dougmsft replied:
We need to have at least one message as part of the commit to verify the entire process works.

markusweimer commented on lang/cs/Org.Apache.REEF.Bridge.CLR/Org.Apache.REEF.Bridge.CLR.csproj:
```diff
@@ -0,0 +1,199 @@
+  <Import Project="$(MSBuildToolsPath)\Microsoft.CSharp.targets" />
+  <Import Project="$(SolutionDir)\.nuget\NuGet.targets" Condition="Exists('$(SolutionDir)\.nuget\NuGet.targets')" />
+  <Import Project="$(PackagesDir)\StyleCop.MSBuild.$(StyleCopVersion)\build\StyleCop.MSBuild.Targets" Condition="Exists('$(PackagesDir)\StyleCop.MSBuild.$(StyleCopVersion)\build\StyleCop.MSBuild.Targets')" />
+  <UsingTask TaskName="AvscToCSharp" TaskFactory="CodeTaskFactory" AssemblyFile="$(MSBuildToolsPath)\Microsoft.Build.Tasks.v4.0.dll">
```
Same comment as above: Consider moving this to its own `.tasks` file.

dougmsft replied:
Okay

tcNickolas commented:
This change doesn't build; all three builds in AppVeyor fail with the same error:

         C:\projects\reef\lang\cs\Org.Apache.REEF.Bridge.CLR\Org.Apache.REEF.Bridge.CLR.csproj(191,5): error MSB3030: Could not copy the file "..\packages\Microsoft.Avro.Core.0.1.0\lib\net451\Microsoft.Avro.Core.dll" because it was not found.

Is there something wrong with the way dependencies are set? Please investigate.

dougmsft commented:
Investigating.

dougmsft commented:
When I build locally from the command prompt with only VS 2017 (MSBuild 15.0), it fails due to a check in the .nuget directory that looks at the OS version and thinks it is running on WindowsNT. This code was added to fix a bug. Do we still need it?

Task "Exec" skipped, due to false condition; ('$(OS)' != 'Windows_NT' And Exists('$(PackagesConfig)')) was evaluated as ('Windows_NT' != 'Windows_NT' And Exists('E:
\reef\lang\cs\Org.Apache.REEF.Tang\packages.config')).
Using "Exec" task from assembly "Microsoft.Build.Tasks.Core, Version=15.1.0.0, Culture=neutral, PublicKeyToken=b03f5f7f11d50a3a".
Task "Exec"
  "E:\reef\lang\cs\.nuget\NuGet.exe" install "E:\reef\lang\cs\Org.Apache.REEF.Tang\packages.config" -source ""  -NonInteractive -RequireConsent -solutionDir "E:\ree
  f\lang\cs\ "
  Restoring NuGet packages...
  To prevent NuGet from downloading packages during build, open the Visual Studio Options dialog, click on the Package Manager node and uncheck 'Allow NuGet to down
  load missing packages'.
E:\reef\lang\cs\.nuget\NuGet.targets(135,9): error : Unable to find version '1.5.6' of package 'Microsoft.Hadoop.Avro'. [E:\reef\lang\cs\Org.Apache.REEF.Tang\Org.Ap
ache.REEF.Tang.csproj]
E:\reef\lang\cs\.nuget\NuGet.targets(135,9): error : Unable to find version '8.0.3' of package 'Newtonsoft.Json'. [E:\reef\lang\cs\Org.Apache.REEF.Tang\Org.Apache.R
EEF.Tang.csproj]
E:\reef\lang\cs\.nuget\NuGet.targets(135,9): error : Unable to find version '2.0.0.668' of package 'protobuf-net'. [E:\reef\lang\cs\Org.Apache.REEF.Tang\Org.Apache.
REEF.Tang.csproj]
E:\reef\lang\cs\.nuget\NuGet.targets(135,9): error : Unable to find version '4.7.49.1' of package 'StyleCop.MSBuild'. [E:\reef\lang\cs\Org.Apache.REEF.Tang\Org.Apac
he.REEF.Tang.csproj]
E:\reef\lang\cs\.nuget\NuGet.targets(135,9): error MSB3073: The command ""E:\reef\lang\cs\.nuget\NuGet.exe" install "E:\reef\lang\cs\Org.Apache.REEF.Tang\packages.c
onfig" -source ""  -NonInteractive -RequireConsent -solutionDir "E:\reef\lang\cs\ "" exited with code 1. [E:\reef\lang\cs\Org.Apache.REEF.Tang\Org.Apache.REEF.Tang.
csproj]
Done executing task "Exec" -- FAILED.
Done building target "RestorePackages" in project "Org.Apache.REEF.Tang.csproj" -- FAILED.
Done Building Project "E:\reef\lang\cs\Org.Apache.REEF.Tang\Org.Apache.REEF.Tang.csproj" (default targets) -- FAILED.

dougmsft commented:
I was able to fix the msbuild 15.0 command line build by adding nuget.org to the nuget.config file. Now I can replicate and debug the AppVeyor build issue.

singlis commented on ':
```diff
@@ -0,0 +1,178 @@
+    <Reference Include="System.Drawing" />
+    <Reference Include="System.IO.Compression.FileSystem" />
+    <Reference Include="System.Numerics" />
+    <Reference Include="System.Reactive.Core, Version=2.2.5.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL" />
```
Both Reactive libraries have a version defined in build.props, so I would change Version=2.2... to Version=$(RxVersion)

dougmsft replied:
removed

singlis commented on ':
```diff
@@ -0,0 +1,178 @@
+      <Using Namespace="System.IO" />
+      <Using Namespace="System.Diagnostics" />
+      <Code Type="Fragment" Language="cs"><![CDATA[
+        // Create a temporary working directory.
```
I think this could use some high level description on what this code is doing.

dougmsft replied:
done

singlis commented on ':
```diff
@@ -0,0 +1,178 @@
+      <HintPath>..\packages\Microsoft.Avro.Tools.0.1.0\lib\net451\Microsoft.Avro.Tools.exe</HintPath>
+    </Reference>
+    <Reference Include="Microsoft.CSharp" />
+    <Reference Include="Microsoft.Hadoop.Avro, Version=1.5.6.0, Culture=neutral, PublicKeyToken=31bf3856ad364e35, processorArchitecture=MSIL">
```
Can you use $(AvroVersion) here?

dougmsft replied:
Yes

dougmsft commented:
I addressed all of the comments except for move the build tasks into separate task files, but I am going to pull and resubmit comments are out of sync even though I did not rebase.
